graphlearn/examples/tf/trainer.py

```diff
@@ -39,6 +40,95 @@
 run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
 run_metadata = tf.RunMetadata()
 
+
+def meta_path_sample(ego, node_type, edge_type, ego_name, nbrs_num, sampler):
+    """ creates the meta-math sampler of the input ego.
+    Args:
+    ego: A query object, the input centric nodes/edges
+    ego_type: A string, the type of `ego`, node_type such as 'paper' or 'user'.
+    ego_name: A string, the name of `ego`.
+    nbrs_num: A list, the number of neighbors for each hop.
+    sampler: A string, the strategy of neighbor sampling.
+    """
+    meta_path = []
+    hops = range(len(nbrs_num))
+    meta_path = ['outV' for i in hops]
+    alias_list = [ego_name + '_hop_' + str(i + 1) for i in hops]
+    mata_path_string = ""
+    for path, nbr_count, alias in zip(meta_path, nbrs_num, alias_list):
+        mata_path_string += path + '(' + edge_type + ').'
+        ego = getattr(ego, path)(edge_type).sample(nbr_count).by(sampler).alias(alias)
+    print("Sampling meta path for {} is {}.".format(node_type, mata_path_string))
+    return ego
+
+def node_embedding(graph, model, node_type, edge_type, **kwargs):
+    """ save node embedding.
+
+    Args:
+    node_type: such as 'paper' or 'user'. 
+    edge_type: such as 'node_type' or 'edge_type'.
+    Return:
+    iterator, ids, embedding.
+    """
+    tfg.conf.training = False
+    ego_name = 'save_node_' + node_type
+    seed = graph.V(node_type).batch(kwargs.get('batch_size', 64)).alias(ego_name)
+    query_save = meta_path_sample(seed, node_type, edge_type, ego_name, kwargs.get('nbrs_num', [10, 5]), kwargs.get('sampler', 'random_without_replacement')).values()
+    dataset = tfg.Dataset(query_save, window=kwargs.get('window', 10))
+    ego_graph = dataset.get_egograph(ego_name)
+    emb = model.forward(ego_graph)
+    return dataset.iterator, ego_graph.src.ids, emb
+
+def dump_embedding(sess, save_iter, save_ids, save_emb, save_file, block_max_lines):
```

Comment:
There is a similar function in https://github.com/alibaba/graph-learn/blob/master/graphlearn/examples/tf/trainer.py#L212, can we just keep one of them?

Reply:
> There is a similar function in https://github.com/alibaba/graph-learn/blob/master/graphlearn/examples/tf/trainer.py#L212, can we just keep one of them?

hello, thanks for you reply. I refactor the codes for a more clean code.  There is no essential update in these codes, all methods are hidden in the source code(in other places or functions), but it is still useful for users, especially users (especially novices) who use unsupervised GraphSage, they don't know how to save the embedding vector at all.

Reply:
Thanks for your contribution!  Just a few more comments :)